src/awkward1/nplike.py

```diff
@@ -269,6 +303,25 @@ def ma(self):
             + awkward1._util.exception_suffix(__file__)
         )
 
+    @property
+    def ndarray(self):
+        return self._module.ndarray
+
+    def asarray(self, array, dtype=None):
+        if isinstance(array, (
+            awkward1.highlevel.Array,
+            awkward1.highlevel.Record,
+            awkward1.layout.Content,
+            awkward1.layout.Record,
+        )):
+            out = awkward1.operations.convert.to_cupy(array)
+            if dtype is not None and out.dtype != dtype:
+                return self._module.asarray(out, dtype=dtype)
+            else:
+                return out
+        else:
+            return self._module.asarray(array, dtype=dtype)
```

Comment:
This is what I was talking about on Friday. CuPy's `asarray` expects the input to be in main memory, but we want it to be like `numpy.asarray` except expect the input to be on the GPU (to get a lot of already-written code to work with a substitution of libraries). The NumPy and CuPy libraries aren't symmetric in that particular way, so we have these classes to act as indirection between what the libraries actually do and what we need them to do. We need `asarray` to be `to_cupy` for GPU-bound Awkward arrays, so I just overloaded that behavior here. There might be a few more subtleties, but we can fix them as they come up because we have this indirection layer.